single-registry-architecture/README.md

```diff
@@ -0,0 +1,217 @@
+Check the discussion [below](#open-questions).
+
+### Choosing OCI repository provider
+
+We checked for the single best registry and chose ACR. We did some evaluation before and you can read results here:
+
+- <https://github.com/giantswarm/roadmap/issues/2382>
+- <https://hackmd.io/gqF6PT53RcG_i4OhrmtWIw#Note-on-compatibility>
+
+Additionally, ACR offers th following:
+
+- High availability by definition.
+- Automatic geo-replication, including geo-replication between Europe and China (confirmed by Azure support, requires
+  going through a formal process, though).
+- Automatic retention
+  - When enabled, manifests that don't have any associated tags (untagged manifests) and are not locked, will be
+    automatically deleted after the number of retention days specified. [Docs](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-retention-policy).
+- Best-in-class support for the OCI standard (it's MS that is pushing it forward).
+- Reasonable pricing
+  - price per region €1.540/day (500 GB included), then €0.003/GB/day + traffic fee
+  - [Azure cost calculator](https://azure.microsoft.com/en-us/pricing/calculator/?service=container-registry)
+
+### Choosing caching solution
+
+There are at least a few solutions possible, below is a list of solutions evaluated:
+
+- <https://goharbor.io>
+  - Seems very complex to run, just [the list of needed components](https://goharbor.io/docs/2.2.0/install-config/) scares
+    me off. We want a simple cache, not all the bells and whistles with a ton of components we have to keep running.
+- <https://github.com/enix/kube-image-keeper>
+  - seems to support container images only, works by a webhook rewriting pod's image - doesn't meet assumptions.
+- <https://gitlab.cronce.io/foss/oci-registry>
+  - OCI compliant, works with `containerd` by registering as a mirror, no webhooks, optional S3 storage - seems
+    like exactly what w need.
+  - Tested, I unfortunately [couldn't make it work](https://github.com/mcronce/oci-registry/issues/14) even for a simple
+  - test case, doesn't seem to be well supported.
+- [ACR connected registry](https://learn.microsoft.com/en-us/azure/container-registry/intro-connected-registry)
+        - there's no mention of how to deploy outside of AKS edge cluster, seems Azure IoT edge thing only
+- [docker's distribution/distribution](https://github.com/distribution/distribution)
+  - one instance can proxy only for a single upstream registry (but that's OK for us)
+  - tested, works with `containerd` for single upstream repo, works as well with `helm` charts
+  - no extra dependencies, can work with just local filesystem storage
+  - exposes reasonable prometheus metrics (transfer times, cache hit ratio)
+  - <https://github.com/distribution/distribution/blob/main/docs/configuration.md#proxy>
+  - <https://github.com/docker/docs/blob/main/content/registry/recipes/mirror.md>
+
+As a result, it seems we can use the `distribution` project from docker.
```

Comment:
I think there is more to consider here - `distribution` has no user management / SSO, no storage possibilities apart from local filesystem.

Additionally AFAIK deployment of `distribution` is essentially "figure it out yourself", there is no official guidance apart from "here are some scripts to run it on your local machine". There are some guides online, but it would be something that we need to take care of.

Reply:
`distribution` in this case is for caching purposes only.  
The user management should be done in the upstream registry and not at the cache level. So personally I see this as a plus.

Regarding the storing capabilities, considering this is a cache, I personally think it's fine. The only consequence is that only one replicas can run at a given point in time. Also from time to time, the mounted volume should be cleaned up, because I doubt the cache checks with upstream whether all the blobs are still present. So this operation is as simple as deleting the volume and creating a new one.

What I would be more interested in finding out is how it manages changes in the upstream manifest and whether it always checks the upstream manifest before serving the cached content.
This is important, for example, in case customers use the `latest` tag for development purposes and always override the `latest` image. The cache needs to know about it, otherwise the users might get unexpected results

Reply:
OK, commenting one-by-one about `dsitribution`

1. Running is pretty easy, I already tested it in a local-storage mode, just works (TM)
2. Storage has options, it can be either local or S3 ([reference](https://github.com/docker/docs/blob/main/content/registry/deploying.md#customize-the-storage-back-end), not tested. S3 obviously makes it easy, but we have to be cross-provider, including on-prem, and here's where the dragons start. As such, I'm leaning towards filesystem storage, as it's easily supported everywhere - but has its own drawbacks (scaling size). Please note, that it's a read-only cache. And that it supports TTL for items, so it's not just super dumb save-and-forget solution.
3. It can run in HA mode, but then it either needs S3 storage or each Pod has its own copy of the cached data (so this increases the cost and miss-rate). [ref](https://github.com/docker/docs/blob/main/content/registry/deploying.md#run-the-registry-as-a-service)
4. `latest` tag is very good point, I'm not 100% sure, but I can easily test that. But I'm 90% certain, that when running in the debug mode, it was always doing HEAD upstream to check for SHA. A question is 'so what's when upstream is down and doens't respond'.

Reply:
One more comment: please not, that even if we run the registry as a single Pod, it's not a SPoF, as we want to configure all the client services with explicit fail over to upstream registry. This obviously means higher cost and probably higher latency, but at least is not critical.